src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c

```diff
@@ -1783,6 +1906,17 @@ static void glb_update_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *policy,
         &glb_policy->lb_channel_connectivity,
         &glb_policy->lb_channel_on_connectivity_changed, NULL);
   }
+
+  // Propagate update to \a fallback_backend_addresses if the fallback is still 
+  // necessary, which is indicated by a non-NULL \a fallback_backend_addresses.
+  if (glb_policy->fallback_backend_addresses != NULL) {
+    grpc_lb_addresses_destroy(exec_ctx, glb_policy->fallback_backend_addresses);
```

Comment:
If the resolver removes all backend addresses, that means that the backends have been removed, and we do not want to keep using the old values.  This is what would happen if we were actually using the round_robin policy directly, and I don't see why the behavior should be any different here.  So I think the current code is fine.

I'm not sure why we deliberately skip serverlists with zero entries, since that should never happen.  Maybe @dgquintas knows.

Reply:
Why should the `serverlist` never have zero entries? I am also curious why we don't want to accept the balancer's advice when it tells us that the `serverlist` is empty. My best guess is that we think having something to try is better than nothing.

Reply:
Thinking about this further, I suspect that this was originally added to deal with the fact that the initial response from the balancer will not include a serverlist.  I cleaned up the way that we handle the initial response from the balancer in #10821, but I didn't remove this case, since I didn't know if it was in use for anything else.  But in retrospect, that was probably too conservative of me, since the balancer protocol clearly shows that all subsequent responses should include a serverlist.

@dgquintas, can you confirm that this check was originally added just to deal with the initial balancer response not including a serverlist?  If so, we can probably just remove the check.

Reply:
Load balancers routinely return empty serverlists upon (re)starting, when they still having received information about backends. After asking @summerxyt (Yetian) why don't the LBs wait to send a non-empty response instead, he mentioned it's just how the current (internal) LB server implementation behaves. 

@markdroth it's not clear which check you are referring to, this thread is attached to new code. Perhaps [this](https://github.com/grpc/grpc/blob/9811915ba3fa1ccdf44b6a70fe1b1dd4782cd508/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1619)? 

About what to do when we receive an empty serverlist after having received non-empty ones: we do currently use the old one. Discarding it would be as simple as making `glb_policy->serverlist = NULL` [after this line](https://github.com/grpc/grpc/blob/9811915ba3fa1ccdf44b6a70fe1b1dd4782cd508/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1623). In that case, following picks would block, the same way they do when all we've received are empty ones. What to do is pretty much up to us. @slash-lib , any thoughts?

Reply:
Yes, that's the check I was asking about.

If the only time that we ever get empty balancer lists is upon balancer startup, then we won't ever get an empty serverlist after we've already gotten a non-empty serverlist.  In that case, the check is probably fine as-is.

@slash-lib can correct me if I'm wrong, but if we did ever somehow get an empty serverlist after getting a non-empty one, I don't think we want to throw out the existing serverlist.

Reply:
Yetian (whom I've summoned to write his own comment on this thread) has confirmed that "we should keep use the last cached server list." in that scenario. I'll let him expand on this.

Reply:
The original question is related to  fallback backend resolution. In the current pending implementation, the resolver will never send a "delete all" (or say empty list) update. But the implementation is subject to change based on review.

Same argument is applied on the balancer resolution.

As for the server list returned from balancer to client, the current implementation can return an empty server list at ANY time. And yes, please keep using the old non-empty list if you receive an empty list.